tx-pool/src/component/pool_map.rs

```diff
@@ -0,0 +1,544 @@
+                        let rejects = std::iter::repeat(reject).take(entries.len());
+                        conflicts.extend(entries.into_iter().zip(rejects));
+                    }
+                }
+            }
+        }
+
+        conflicts
+    }
+
+    // fill proposal txs
+    pub(crate) fn fill_proposals(
+        &self,
+        limit: usize,
+        exclusion: &HashSet<ProposalShortId>,
+        proposals: &mut HashSet<ProposalShortId>,
+        status: &Status,
+    ) {
+        for entry in self.score_sorted_iter_by(vec![*status]) {
+            if proposals.len() == limit {
+                break;
+            }
+            let id = entry.proposal_short_id();
+            if !exclusion.contains(&id) {
+                proposals.insert(id);
+            }
+        }
+    }
+
+    pub(crate) fn iter(&self) -> impl Iterator<Item = &PoolEntry> {
+        self.entries.iter().map(|(_, entry)| entry)
+    }
+
+    pub(crate) fn next_evict_entry(&self, status: Status) -> Option<ProposalShortId> {
+        self.entries
+            .iter_by_evict_key()
+            .find(move |entry| entry.status == status)
+            .map(|entry| entry.id.clone())
+    }
+
+    pub(crate) fn clear(&mut self) {
+        self.entries = MultiIndexPoolEntryMap::default();
+        self.edges.clear();
+        self.links.clear();
+    }
+
+    pub(crate) fn score_sorted_iter_by(
+        &self,
+        statuses: Vec<Status>,
+    ) -> impl Iterator<Item = &TxEntry> {
+        self.entries
+            .iter_by_score()
+            .rev()
+            .filter(move |entry| statuses.contains(&entry.status))
+            .map(|entry| &entry.inner)
+    }
+
+    fn remove_entry_links(&mut self, id: &ProposalShortId) {
+        if let Some(parents) = self.links.get_parents(id).cloned() {
+            for parent in parents {
+                self.links.remove_child(&parent, id);
+            }
+        }
+        if let Some(children) = self.links.get_children(id).cloned() {
+            for child in children {
+                self.links.remove_parent(&child, id);
+            }
+        }
+        self.links.remove(id);
+    }
+
+    fn update_ancestors_index_key(&mut self, child: &TxEntry, op: EntryOp) {
+        let ancestors: HashSet<ProposalShortId> =
+            self.links.calc_ancestors(&child.proposal_short_id());
+        for anc_id in &ancestors {
+            // update parent score
+            let entry = self.entries.get_by_id(anc_id).unwrap().clone();
+            let mut parent = entry.inner.clone();
+            match op {
+                EntryOp::Remove => parent.sub_descendant_weight(child),
+                EntryOp::Add => parent.add_descendant_weight(child),
+            }
+            let short_id = parent.proposal_short_id();
+            self.entries.modify_by_id(&short_id, |e| {
+                e.evict_key = parent.as_evict_key();
+                e.inner = parent;
+            });
```

Comment:
can we use `modify_by_id` directly here?
```suggestion
            // update parent score
            self.entries.modify_by_id(anc_id, |e| {
                match op {
                    EntryOp::Remove => e.inner.sub_descendant_weight(child),
                    EntryOp::Add => e.inner.add_descendant_weight(child),
                };
                e.evict_key = e.inner.as_evict_key();
            });
```